incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +59,306 @@ type ObjectReconcilerNew struct {
 
 	// GVK is the group/version/kind handled by this reconciler.
 	GVK schema.GroupVersionKind
+
+	// Affected is a channel of event.GenericEvent (see "Watching Channels" in
+	// https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html) that is used to
+	// enqueue additional objects that need updating.
+	Affected chan event.GenericEvent
 }
 
 // +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;create;update;patch;delete
 
 func (r *ObjectReconcilerNew) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	resp := ctrl.Result{}
+	ctx := context.Background()
 	log := r.Log.WithValues("trigger", req.NamespacedName)
-	log.Info("New object controller placeholder")
-	return resp, nil
+
+	log.Info("Reconciling object")
+
+	// Read the object. If it's missing, assume it's been deleted or not yet created.
+	inst := &unstructured.Unstructured{}
+	inst.SetGroupVersionKind(r.GVK)
+	inst.SetNamespace(req.Namespace)
+	inst.SetName(req.Name)
+	found := true
+	if err := r.Get(ctx, req.NamespacedName, inst); err != nil {
+		if !errors.IsNotFound(err) {
+			log.Info("Couldn't read")
+			return resp, err
+		}
+		found = false
+	}
+	act, so := r.syncWithForest(ctx, log, found, inst)
+
+	return resp, r.operate(ctx, log, act, inst, so)
+}
+
+// syncWithForest syncs the object instance with the in-memory forest. This is the only place we should
+// lock the forest in each Reconcile, so this fn needs to return everything relevant for the rest of the
+// Reconcile. This fn shouldn't contact the apiserver since that's a slow operation and everything will
+// block on the lock being held. It returns the action to take on the object (delete, overwrite or do
+// nothing) and a source object if the action is to overwrite it.
+func (r *ObjectReconcilerNew) syncWithForest(ctx context.Context, log logr.Logger, found bool, inst *unstructured.Unstructured) (action, *unstructured.Unstructured) {
+	r.Forest.Lock()
+	defer r.Forest.Unlock()
+
+	if found {
+		// If it's about to be deleted, do nothing, just wait for it to be actually deleted.
+		if !inst.GetDeletionTimestamp().IsZero() {
+			log.Info("Will soon be deleted")
+			return unchanging, nil
+		}
+		return r.syncObject(ctx, log, inst)
```

Comment:
Nit: You can combine this with line 127, and just set the label if it's not found.

This would suggest a high-level function that looks something like:

```
if !inst.GetDeletionTimestamp().IsZero() {
  return ignore, nil
}

if inst.GetCreationTimestamp().IsZero() {
  // obj doesn't exist
  if r.isSource(inst) {
    r.removeDeletedSource(ctx, log, inst) // this can also call enqueueDescendants
    return ignore, nil
  }
  
  // This object should be propagated
  metadata.SetLabel(inst, api.LabelInheritedFrom, "sns placeholder")
}

return r.syncObject(ctx, log, inst)
```

I personally find this a bit more readable - the "weird" cases are dealt with up-front, and then we "fall through" to the regular case. But that's up to you.